onboard/merge_iterator.go

```diff
@@ -0,0 +1,170 @@
+package onboard
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/treeverse/lakefs/catalog/rocks"
+)
+
+// prefixMergeIterator takes inventory iterator and commit iterator and merges them.
+// If an entry's prefix is in `prefixes`, it would be taken from invIt.
+// No entry from committedIt would be returned if its prefix is in prefixes.
+type prefixMergeIterator struct {
+	invIt       rocks.EntryIterator
+	committedIt rocks.EntryIterator
+
+	err      error
+	bothDone bool
+}
+
+func newPrefixMergeIterator(iterator rocks.EntryIterator, it rocks.EntryListingIterator, prefixes []string) rocks.EntryIterator {
+	return &prefixMergeIterator{
+		invIt:       iterator,
+		committedIt: newIgnorePrefixIterator(it, prefixes),
+		err:         nil,
+		bothDone:    false,
+	}
+}
+
+func (pmi *prefixMergeIterator) Next() bool {
+	if pmi.err != nil {
+		return false
+	}
+
+	switch {
+	case pmi.invIt.Value() == nil && pmi.committedIt.Value() == nil:
+		pmi.bothDone = true
+		return false
+
+	case pmi.invIt.Value() == nil:
+		return pmi.advanceIt(pmi.committedIt)
+
+	case pmi.committedIt.Value() == nil:
+		return pmi.advanceIt(pmi.invIt)
+
+		// from now on - both iterator has values,
+		// so we'll always return true
+	case strings.Compare(pmi.committedIt.Value().Path.String(), pmi.invIt.Value().Path.String()) < 0:
+		_ = pmi.advanceIt(pmi.committedIt)
+		return true
+
+	default:
+		_ = pmi.advanceIt(pmi.invIt)
+		return true
+	}
+}
+
+func (pmi *prefixMergeIterator) advanceIt(it rocks.EntryIterator) bool {
```

Comment:
`it` is copied by value

Reply:
You sure about that? `it` is an interface.. I guess tests will tell :)

Reply:
my bad - missed that part it is an interface.